File: src/main/java/com/burgstaller/okhttp/CachingAuthenticatorDecorator.java
package com.burgstaller.okhttp;

import java.io.IOException;
import java.util.Map;

import com.burgstaller.okhttp.digest.CachingAuthenticator;

import okhttp3.Authenticator;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.Route;

/**
 * An authenticator decorator which saves the generated authentication headers for a specific host.
 * To be used in tandem with {@link AuthenticationCacheInterceptor}.
 * Depending on your use case you will probably need to use a {@link java.util.concurrent.ConcurrentHashMap}.
 */
public class CachingAuthenticatorDecorator implements Authenticator {
    private final Authenticator innerAuthenticator;
    private final Map<String, CachingAuthenticator> authCache;
    private final CacheKeyProvider cacheKeyProvider;

    public CachingAuthenticatorDecorator(Authenticator innerAuthenticator, Map<String, CachingAuthenticator> authCache, CacheKeyProvider cacheKeyProvider) {
        this.innerAuthenticator = innerAuthenticator;
        this.authCache = authCache;
        this.cacheKeyProvider = cacheKeyProvider;
    }

    public CachingAuthenticatorDecorator(Authenticator innerAuthenticator, Map<String, CachingAuthenticator> authCache,boolean proxy) {
        this(innerAuthenticator, authCache, proxy?new DefaultProxyCacheKeyProvider():new DefaultRequestCacheKeyProvider());
    }

    public CachingAuthenticatorDecorator(Authenticator innerAuthenticator, Map<String, CachingAuthenticator> authCache) {
        this(innerAuthenticator, authCache, false);
    }
    @Override
    public Request authenticate(Route route, Response response) throws IOException {
        Request authenticated = innerAuthenticator.authenticate(route, response);
        if (authenticated != null) {
            String authorizationValue;
            if(cacheKeyProvider.applyToProxy()){
                authorizationValue = authenticated.header("Proxy-Authorization");
            }else{
                authorizationValue = authenticated.header("Authorization");
            }

            if (authorizationValue != null && innerAuthenticator instanceof CachingAuthenticator) {
                String key;
                if(cacheKeyProvider.applyToProxy()){
                    key = cacheKeyProvider.getCachingKey(route.proxy());
                }else {
                    key = cacheKeyProvider.getCachingKey(authenticated);
                }
                authCache.put(key, (CachingAuthenticator) innerAuthenticator);
            }
        }
        return authenticated;
    }
}

File: src/main/kotlin/com/burgstaller/okhttp/CachingAuthenticatorDecorator.kt
package com.burgstaller.okhttp

import com.burgstaller.okhttp.digest.CachingAuthenticator
import okhttp3.Authenticator
import okhttp3.Request
import okhttp3.Response
import okhttp3.Route

/**
 * An authenticator decorator which saves the generated authentication headers for a specific host.
 * To be used in tandem with [AuthenticationCacheInterceptor].
 * Depending on your use case you will probably need to use a [java.util.concurrent.ConcurrentHashMap].
 */
class CachingAuthenticatorDecorator(private val innerAuthenticator: Authenticator, private val authCache: MutableMap<String, CachingAuthenticator>) : Authenticator {
  override fun authenticate(route: Route?, response: Response): Request? {
    val authenticated = innerAuthenticator.authenticate(route, response)
    if (authenticated != null) {
      val authorizationValue = authenticated.header("Authorization")
      if (authorizationValue != null && innerAuthenticator is CachingAuthenticator) {
        val url = authenticated.url()
        val key = CachingUtils.defaultCacheKey(url)
        authCache[key] = innerAuthenticator
      }
    }
    return authenticated
  }
}
